### PR TITLE
Full CI support for public builds + switch to use cibuildwheel

### LINKED ISSUE #227: Validate wheel build in CI/CD
We need to build wheels in the public CI because
- it generates a log that can be audited by non-team members when an issue happens (ex: #226)
- the artifacts can be inspected and reused (ex: moving to internal artifactory for QA/release processes)
- wheel contents can be tested as part of CI

### PR DESCRIPTION
**Pull request by leofang**
Close #227. Part of #81.

The earlier build pipeline was deficient because
1. it only covered a single point in our large support matrix (ex: no aarch64/win-64 coverage)
2. it used a miniforge container to build wheels, which does not guarantee compatibility with the manylinux/glibc requirement
3. since a container is used, it is impossible to reuse many widely used/validated, OSS GHA workflows, and everything has to be homegrown scripts, which is an unnecessary burden to both developer and devops teams
   - one addition downside of this is it's really hard to understand the overall CI design, since codes are scattered into many files that are hard to follow
4. the old CI had too many unnecessary pieces not suitable for this project (ex: CMake, Legate, ...)

This PR switches to use public GHA workflows. Most steps are meant to be easily reusable (after some tweaking/refactoring) by other Python/CUDA projects, either first-party or third-party. For example, `cibuildwheels` has been used to offer guarantee to manylinux. It also lays the ground work for later bring-up of the CI test stage. (This PR only covers the build stage.) 

As of commit https://github.com/NVIDIA/cuda-python/commit/0d554f0945b1bd1b7797170ece2a13765ab6dbe8, we can build all needed wheels using GHA now! 🎉
- `cuda.bindings`
- `cuda.core`
- Python: 3.9-3.12
- OS: linux-64, linux-aarch64, win-64
- CUDA 12.6 (we can evaluate backporting to the cuda-11 branch)
 
The artifacts are available here for inspection: https://github.com/NVIDIA/cuda-python/actions/runs/12192916020

### CODE REVIEW

**Comment by ksimpson-work** [comment]
I'm really liking the look of this so far :)

**Comment by leofang** [comment]
OK  we can build on all platforms now! 🎉

**Comment by leofang** [comment]
OK, `cuda.core` is built successfully for all supported Python x OS versions, the artifacts can be retrieved/inspected here:
https://github.com/NVIDIA/cuda-python/actions/runs/12189979543
I am going to work on building `cuda.bindings` (for CUDA 12) next.

**Comment by leofang** [comment]
As of commit 0d554f0, we can build all needed wheels (`cuda.{bindings,core}` x Python 3.9-3.12 x {linux-64, linux-aarch64, win-64} x CUDA 12.x) using GHA now! 🎉 Artifacts are available here:
https://github.com/NVIDIA/cuda-python/actions/runs/12192916020

**Comment by leofang** [comment]
(updated the PR description)

File: .github/actions/build/action.yml
Comment (leofang) [inline .github/actions/build/action.yml:66]: Note: This was commented out because of #271.

File: .github/actions/setup/action.yml
Comment (leofang) [inline .github/actions/setup/action.yml:39]: Technically we don't need this step. However, because `actions/setup-python` is not relocatable (its RPATH is hard-wired, see https://github.com/actions/setup-python/issues/871), on a self-hosted runner cibuildwheels cannot properly load libpython3.12. This is a WAR by forcing the same Python version installed on the host (and mounted into the container).

File: .github/actions/setup/action.yml
Comment (leofang) [inline .github/actions/setup/action.yml]: We can clean this up now, as it's now tracked in https://github.com/nv-gha-runners/vm-images/issues/102. (Our images do not contain pre-installed `pwsh`.)

File: .github/actions/setup/action.yml
Comment (leofang) [inline .github/actions/setup/action.yml]: We can clean this up now, as it's now tracked in https://github.com/nv-gha-runners/vm-images/issues/103. (Our images do not contain pre-installed MSVC.)

File: .github/actions/setup/action.yml
Comment (keenan-simpson) [inline .github/actions/setup/action.yml:115]: nit: we could format the python version instead of doing the if else fallthrough:

```yml
        CIBW_BUILD='cp$(echo '${{ inputs.python-version }}' | tr -d '.')-manylinux*'

        CIBW_BUILD='cp$(echo '${{ inputs.python-version }}' | tr -d '.')-win_amd64*'
```

Comment (ksimpson-work) [inline .github/actions/setup/action.yml:157]:
```suggestion
        CIBW_BUILD='cp{$(echo '${{ inputs.python-version }}' | tr -d '.')}-manylinux*'
```

File: .github/workflows/gh-build-and-test.yml
Comment (keenan-simpson) [inline .github/workflows/gh-build-and-test.yml:4]: This could be saved for the next PR, but having this called build-and-test has been misleading imo. I think it could be better to call gh-build.yml and gh-test.yml from ci-gh.yml

File: .github/actions/setup/action.yml
Comment (keenan-simpson) [inline .github/actions/setup/action.yml]: My gut tells me we should have the windows stuff uncommented before approval, or take all windows related changes out of this review. Having the install-vs script included with this commented out for example. Per your recent comments it seems this is being address by the nv-gha-runnners issues. Do we want to consider those dependencies?

Comment (keenan-simpson) [reply .github/actions/setup/action.yml]: nvm per plan to switch to nv pre installed runners. Will wait to approve until that stuff is removed?

File: .github/workflows/gh-build-and-test.yml
Comment (leofang) [reply .github/workflows/gh-build-and-test.yml:4]: Discussed on the call, yes we'll have to refactor this to accommodate that we have different build/test matrices to cover.

File: .github/actions/setup/action.yml
Comment (leofang) [reply .github/actions/setup/action.yml]: - Using GH-hosted runners: pwsh and msvc are both included in the Windows images.
- Using self-hosted (nv-gha) runners: Yes we need them to support both, this is currently a blocker tracked in both vm-images issues.

File: .github/actions/setup/action.yml
Comment (leofang) [reply .github/actions/setup/action.yml]: removed

File: .github/actions/setup/action.yml
Comment (leofang) [reply .github/actions/setup/action.yml]: removed in [19445d5](https://github.com/NVIDIA/cuda-python/pull/267/commits/19445d58143f0dc58193a39080c5808a4e52d220)

**Comment by ksimpson-work** [comment]
Last comment I think: Perhaps we should squash these commits. Not hellbent on this

File: .github/actions/setup/action.yml
Comment (keenan-simpson) [inline .github/actions/setup/action.yml:62]: or could do it like this with the PYTHON_VERSION_FORMATTED var

File: .github/actions/setup/action.yml
Comment (leofang) [reply .github/actions/setup/action.yml:115]: Done in 07b8d1d.

File: .github/actions/setup/action.yml
Comment (leofang) [reply .github/actions/setup/action.yml:62]: yup, done in commit 07b8d1d

**Comment by leofang** [comment]
> Perhaps we should squash these commits. 

Yes when we merge there's a drop-down menu to select merge or squash.

**Comment by ksimpson-work** [comment]
ahh, that option is grayed out for me on this project

**Review by ksimpson-work — APPROVED** [review]

**Comment by leofang** [comment]
@ksimpson-work is it still grey after you approved it?

**Comment by ksimpson-work** [comment]
No, I noticed it when I went to merge my own PRs. I'll take a screenshot next time, unless I was hallucinating

**Comment by leofang** [comment]
Let's merge! Thanks for quick reviews @ksimpson-work!

**Pull request merged**
